plugins/woocommerce/includes/class-wc-cart.php

```diff
@@ -1810,16 +1810,22 @@ public function is_coupon_emails_allowed( $check_emails, $restrictions ) {
 	 * @return bool
 	 */
 	public function has_discount( $coupon_code = '' ) {
-		return $coupon_code ? in_array(
-			wc_strtolower( wc_format_coupon_code( $coupon_code ) ),
-			array_map(
-				function ( $code ) {
-					return wc_strtolower( wc_format_coupon_code( $code ) );
-				},
-				$this->applied_coupons
-			),
-			true
-		) : count( $this->applied_coupons ) > 0;
+		$applied_coupons = $this->get_applied_coupons();
+
+		if ( ! $coupon_code ) {
+			return count( $applied_coupons ) > 0;
+		}
+
+		$coupon_code = wc_format_coupon_code( $coupon_code );
+
+		// Check if the coupon is in applied coupons using case-insensitive comparison.
+		foreach ( $applied_coupons as $applied_coupon ) {
+			if ( wc_is_same_coupon( $applied_coupon, $coupon_code ) ) {
```

Comment:
should we call `wc_format_coupon_code` on $applied_coupon? The previous code was doing that.

Reply:
I don't think `wc_is_same_coupon` use `wc_format_coupon_code` instead because:
- applied coupons have already had `wc_format_coupon_code` applied
- the pattern is to use `wc_format_coupon_code` on cart codes
- we might end up applying twice the filer to a coupon which might result in different codes